File: windows/MoongateCore.Tests/SubtitleTimingPlannerTests.cs
using Moongate.Core;

namespace MoongateCore.Tests;

public class SubtitleTimingPlannerTests
{
    [Fact]
    public void SpeechAlignedVisibleSeconds_KeepsShortFeedbackBrief()
    {
        Assert.Equal(2.0, SubtitleTimingPlanner.SpeechAlignedVisibleSeconds("Copy.", endsSentence: true));
        Assert.Equal(2.0, SubtitleTimingPlanner.SpeechAlignedVisibleSeconds("What heat?", endsSentence: true));
    }

    [Fact]
    public void SpeechAlignedVisibleSeconds_ExtendsShortEmphaticLines()
    {
        Assert.Equal(2.45, SubtitleTimingPlanner.SpeechAlignedVisibleSeconds("10,000 hours!", endsSentence: true));
        Assert.Equal(2.45, SubtitleTimingPlanner.SpeechAlignedVisibleSeconds("No!!", endsSentence: true));
        Assert.Equal(2.45, SubtitleTimingPlanner.SpeechAlignedVisibleSeconds("Ever again.", endsSentence: true));
        Assert.Equal(1.92, SubtitleTimingPlanner.SpeechAlignedVisibleSeconds("like around week eight,", endsSentence: false));
    }

    [Fact]
    public void SpeechAlignedVisibleSeconds_ScalesLongSpeechWithoutDragging()
    {
        var duration = SubtitleTimingPlanner.SpeechAlignedVisibleSeconds(
            "This is a longer spoken sentence that should stay readable without lingering too long.",
            endsSentence: true);

        Assert.True(duration > 4.0);
        Assert.True(duration <= 9.0);
    }

    [Fact]
    public void WeakSemanticBoundary_DetectsBadEnglishSplits()
    {
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("the", "ship"));
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("moon", "and"));
        Assert.False(SubtitleTimingPlanner.IsWeakBoundary("moon.", "Next"));
    }

    [Fact]
    public void WeakSemanticBoundary_DetectsRomanceLanguageSplits()
    {
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("tono", "de"));
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("de", "ponerte"));
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("des", "chaussures"));
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("di", "fronte"));
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("muchas", "veces"));
        Assert.True(SubtitleTimingPlanner.IsWeakBoundary("no", "hacen"));
    }

    [Fact]
    public void Tokenizers_HandleLatinAndCjkText()
    {
        Assert.Equal(["hello", "starship"], SubtitleTimingPlanner.WordTokens("Hello, Starship!"));
        Assert.Equal(["copy", "10", "engines"], SubtitleTimingPlanner.SpeechTokens("Copy. 10 engines?"));
        Assert.Equal(
            ["sì", "può", "iniziare", "questa", "settimana"],
            SubtitleTimingPlanner.SpeechTokens("Sì, può iniziare questa settimana?"));
        Assert.Equal(["3", "€", "4", "99", "€"], SubtitleTimingPlanner.TimingTokens("3 € 4,99€."));
        Assert.Empty(SubtitleTimingPlanner.SpeechTokens("今天我们继续"));
        Assert.Empty(SubtitleTimingPlanner.SpeechTokens("내가 서 있는 곳"));
        Assert.Equal(7, SubtitleTimingPlanner.VisibleCharacters("你 好 world"));
    }
}

File: Tests/MoongateCoreTests/SubtitleTimingPlannerTests.swift
@testable import MoongateCore
import XCTest

final class SubtitleTimingPlannerTests: XCTestCase {
    func testSpeechAlignedVisibleSecondsKeepsShortFeedbackBrief() {
        XCTAssertEqual(SubtitleTimingPlanner.speechAlignedVisibleSeconds("Copy.", endsSentence: true), 2.0)
        XCTAssertEqual(SubtitleTimingPlanner.speechAlignedVisibleSeconds("What heat?", endsSentence: true), 2.0)
    }

    func testSpeechAlignedVisibleSecondsExtendsShortEmphaticLines() {
        XCTAssertEqual(SubtitleTimingPlanner.speechAlignedVisibleSeconds("10,000 hours!", endsSentence: true), 2.45)
        XCTAssertEqual(SubtitleTimingPlanner.speechAlignedVisibleSeconds("No!!", endsSentence: true), 2.45)
        XCTAssertEqual(SubtitleTimingPlanner.speechAlignedVisibleSeconds("Ever again.", endsSentence: true), 2.45)
        XCTAssertEqual(SubtitleTimingPlanner.speechAlignedVisibleSeconds("like around week eight,", endsSentence: false), 1.92)
    }

    func testSpeechAlignedVisibleSecondsScalesLongSpeechWithoutDragging() {
        let duration = SubtitleTimingPlanner.speechAlignedVisibleSeconds(
            "This is a longer spoken sentence that should stay readable without lingering too long.",
            endsSentence: true
        )

        XCTAssertGreaterThan(duration, 4.0)
        XCTAssertLessThanOrEqual(duration, 9.0)
    }

    func testWeakSemanticBoundaryDetectsBadEnglishSplits() {
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "the", rightToken: "ship"))
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "moon", rightToken: "and"))
        XCTAssertFalse(SubtitleTimingPlanner.isWeakBoundary(leftToken: "moon.", rightToken: "Next"))
    }

    func testWeakSemanticBoundaryDetectsRomanceLanguageSplits() {
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "tono", rightToken: "de"))
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "de", rightToken: "ponerte"))
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "des", rightToken: "chaussures"))
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "di", rightToken: "fronte"))
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "muchas", rightToken: "veces"))
        XCTAssertTrue(SubtitleTimingPlanner.isWeakBoundary(leftToken: "no", rightToken: "hacen"))
    }

    func testTokenizersHandleLatinAndCJKText() {
        XCTAssertEqual(SubtitleTimingPlanner.wordTokens("Hello, Starship!"), ["hello", "starship"])
        XCTAssertEqual(SubtitleTimingPlanner.speechTokens("Copy. 10 engines?"), ["copy", "10", "engines"])
        XCTAssertEqual(
            SubtitleTimingPlanner.speechTokens("Sì, può iniziare questa settimana?"),
            ["sì", "può", "iniziare", "questa", "settimana"]
        )
        XCTAssertEqual(SubtitleTimingPlanner.timingTokens("3 € 4,99€."), ["3", "€", "4", "99", "€"])
        XCTAssertEqual(SubtitleTimingPlanner.speechTokens("今天我们继续"), [])
        XCTAssertEqual(SubtitleTimingPlanner.speechTokens("내가 서 있는 곳"), [])
        XCTAssertEqual(SubtitleTimingPlanner.visibleCharacters("你 好 world"), 7)
    }
}
